File: src/engine/market.ts
import type { Agent, Order, MarketState, TradeEvent } from './types';
import { calculateMRS } from './math';

export const generateOrders = (agents: Agent[]): { bids: Order[]; asks: Order[] } => {
  const bids: Order[] = [];
  const asks: Order[] = [];

  agents.forEach(agent => {
    if (agent.isDead) return;

    const mrs = calculateMRS(agent);
    
    // Safety check for non-finite or 0 MRS
    if (!isFinite(mrs) || isNaN(mrs) || mrs <= 0) return;

    if (mrs > 1.0) {
      // Natural Buyer: Willing to pay up to `mrs` Gold per 1 Food.
      // Max Food they can afford is `goldInventory / mrs`
       const maxQty = Math.floor(agent.goldInventory / mrs);
       if (maxQty > 0) {
         bids.push({
           agentId: agent.id,
           type: 'BID',
           price: mrs,
           qty: maxQty
         });
       }
    } else if (mrs < 1.0) {
       // Natural Seller: Willing to sell 1 Food for at least `mrs` Gold.
       // Max Food they can sell is their entire `foodInventory`
       const maxQty = agent.foodInventory;
       if (maxQty > 0) {
         asks.push({
           agentId: agent.id,
           type: 'ASK',
           price: mrs,
           qty: maxQty
         });
       }
    }
  });

  return { bids, asks };
};

export const resolveMarket = (
  bids: Order[], 
  asks: Order[], 
  agents: Agent[],
  tick: number
): { updatedAgents: Agent[]; marketState: MarketState } => {
  
  // Clone pristine orders to expose the full macroscopic supply/demand curves to the UI
  const pristineBids = bids.map(b => ({ ...b })).sort((a, b) => b.price - a.price);
  const pristineAsks = asks.map(a => ({ ...a })).sort((a, b) => a.price - b.price);

  // Sort Bids descending (highest willingness to pay first)
  const sortedBids = [...bids].sort((a, b) => b.price - a.price);
  // Sort Asks ascending (lowest willingness to accept first)
  const sortedAsks = [...asks].sort((a, b) => a.price - b.price);

  let currentBidIdx = 0;
  let currentAskIdx = 0;

  // We need to mutate agent states sequentially as trades happen
  // Copy the agent map for quick lookup and O(1) replace
  const agentMap = new Map<string, Agent>();
  agents.forEach(a => agentMap.set(a.id, a));

  let lastClearingPrice: number | null = null;
  let volumeLastTick = 0;
  const tradesLastTick: TradeEvent[] = [];

  while (currentBidIdx < sortedBids.length && currentAskIdx < sortedAsks.length) {
    const bid = sortedBids[currentBidIdx];
    const ask = sortedAsks[currentAskIdx];

    if (bid.price >= ask.price && bid.qty > 0 && ask.qty > 0) {
      const matchPrice = (bid.price + ask.price) / 2;
      let matchVolume = Math.min(bid.qty, ask.qty);
      
      const buyer = agentMap.get(bid.agentId);
      const seller = agentMap.get(ask.agentId);
      
      if (!buyer || !seller) {
         if (!buyer) bid.qty = 0;
         if (!seller) ask.qty = 0;
         continue;
      }
      
      // Before executing, enforce hard inventory constraints just in case
      // Buyer has enough gold?
      const maxAffordableVolume = Math.floor(buyer.goldInventory / matchPrice);
      // Seller has enough food?
      const maxSellableVolume = seller.foodInventory;
      
      matchVolume = Math.min(matchVolume, maxAffordableVolume, maxSellableVolume);
      
      if (matchVolume <= 0) {
        // Can't actuate trade: void the constraint
        if (maxAffordableVolume <= 0) bid.qty = 0;
        if (maxSellableVolume <= 0) ask.qty = 0;
        continue;
      }

      // Execute Trade natively!
      agentMap.set(buyer.id, {
        ...buyer,
        goldInventory: buyer.goldInventory - (matchVolume * matchPrice),
        foodInventory: buyer.foodInventory + matchVolume
      });

      agentMap.set(seller.id, {
        ...seller,
        goldInventory: seller.goldInventory + (matchVolume * matchPrice),
        foodInventory: seller.foodInventory - matchVolume
      });

      lastClearingPrice = matchPrice;
      volumeLastTick += matchVolume;
      
      tradesLastTick.push({
        tick,
        buyerId: buyer.id,
        sellerId: seller.id,
        price: matchPrice,
        qty: matchVolume
      });
      
      // Reduce fulfilled order quantities
      bid.qty -= matchVolume;
      ask.qty -= matchVolume;
    } else {
      // Market cleared, spread is positive (bid < ask) and trade is mutually destructive
      break;
    }

    // Advance orders if exhausted
    if (bid.qty <= 0) currentBidIdx++;
    if (ask.qty <= 0) currentAskIdx++;
  }

  // Convert map back to array. Order might have shifted but that's fine. 
  // We can guarantee exact order by mapping original array over the map:
  const updatedAgents = agents.map(a => agentMap.get(a.id)!);

  return {
    updatedAgents,
    marketState: {
      bids: pristineBids,
      asks: pristineAsks,
      lastClearingPrice,
      volumeLastTick,
      tradesLastTick
    }
  };
};

File: econ-sim-elixir/lib/econ_sim/market.ex
defmodule EconSim.Market do
  @moduledoc """
  Market order generation and order book matching.
  Direct port of src/engine/market.ts.
  """

  alias EconSim.Types.{Agent, Order, TradeEvent, MarketState}
  alias EconSim.Math

  @doc """
  Generates buy/sell orders from all living agents based on their MRS.
  Agents with MRS > 1.0 are natural buyers (bid for food with gold).
  Agents with MRS < 1.0 are natural sellers (ask to sell food for gold).
  """
  @spec generate_orders([Agent.t()]) :: %{bids: [Order.t()], asks: [Order.t()]}
  def generate_orders(agents) do
    agents
    |> Enum.filter(&(!&1.is_dead))
    |> Enum.reduce(%{bids: [], asks: []}, fn agent, acc ->
      mrs = Math.calculate_mrs(agent)

      cond do
        not is_number(mrs) or mrs <= 0 ->
          acc

        mrs > 1.0 ->
          max_qty = floor(agent.gold_inventory / mrs)

          if max_qty > 0 do
            order = %Order{agent_id: agent.id, type: :bid, price: mrs, qty: max_qty * 1.0}
            %{acc | bids: [order | acc.bids]}
          else
            acc
          end

        mrs < 1.0 ->
          max_qty = agent.food_inventory

          if max_qty > 0 do
            order = %Order{agent_id: agent.id, type: :ask, price: mrs, qty: max_qty}
            %{acc | asks: [order | acc.asks]}
          else
            acc
          end

        true ->
          acc
      end
    end)
  end

  @doc """
  Resolves the market by matching bids (highest first) against asks (lowest first).
  Returns updated agents and the new market state.
  """
  @spec resolve_market([Order.t()], [Order.t()], [Agent.t()], integer()) ::
          %{updated_agents: [Agent.t()], market_state: MarketState.t()}
  def resolve_market(bids, asks, agents, tick) do
    # Keep pristine copies for UI display
    pristine_bids = bids |> Enum.sort_by(& &1.price, :desc)
    pristine_asks = asks |> Enum.sort_by(& &1.price, :asc)

    sorted_bids = bids |> Enum.sort_by(& &1.price, :desc) |> Enum.map(&Map.from_struct/1)
    sorted_asks = asks |> Enum.sort_by(& &1.price, :asc) |> Enum.map(&Map.from_struct/1)

    agent_map = Map.new(agents, fn a -> {a.id, a} end)

    {agent_map, last_clearing_price, volume, trades} =
      match_orders(sorted_bids, sorted_asks, agent_map, tick, nil, 0, [])

    updated_agents = Enum.map(agents, fn a -> Map.get(agent_map, a.id, a) end)

    %{
      updated_agents: updated_agents,
      market_state: %MarketState{
        bids: pristine_bids,
        asks: pristine_asks,
        last_clearing_price: last_clearing_price,
        volume_last_tick: volume,
        trades_last_tick: Enum.reverse(trades)
      }
    }
  end

  # Recursive order matching — mirrors the while loop from market.ts
  defp match_orders([], _asks, agent_map, _tick, lcp, vol, trades),
    do: {agent_map, lcp, vol, trades}

  defp match_orders(_bids, [], agent_map, _tick, lcp, vol, trades),
    do: {agent_map, lcp, vol, trades}

  defp match_orders(
         [bid | rest_bids] = bids,
         [ask | rest_asks] = asks,
         agent_map,
         tick,
         _lcp,
         vol,
         trades
       ) do
    cond do
      bid.price < ask.price ->
        # Spread is positive — market cleared
        {agent_map, _lcp = if(trades != [], do: List.first(trades).price, else: nil), vol, trades}

      bid.qty <= 0 ->
        match_orders(rest_bids, asks, agent_map, tick, nil, vol, trades)

      ask.qty <= 0 ->
        match_orders(bids, rest_asks, agent_map, tick, nil, vol, trades)

      true ->
        buyer = Map.get(agent_map, bid.agent_id)
        seller = Map.get(agent_map, ask.agent_id)

        if is_nil(buyer) or is_nil(seller) do
          new_bids = if is_nil(buyer), do: [%{bid | qty: 0} | rest_bids], else: bids
          new_asks = if is_nil(seller), do: [%{ask | qty: 0} | rest_asks], else: asks
          match_orders(new_bids, new_asks, agent_map, tick, nil, vol, trades)
        else
          match_price = (bid.price + ask.price) / 2
          match_volume = min(bid.qty, ask.qty)
          max_affordable = floor(buyer.gold_inventory / match_price)
          max_sellable = seller.food_inventory
          match_volume = min(match_volume, min(max_affordable * 1.0, max_sellable))

          if match_volume <= 0 do
            new_bids = if max_affordable <= 0, do: [%{bid | qty: 0} | rest_bids], else: bids
            new_asks = if max_sellable <= 0, do: [%{ask | qty: 0} | rest_asks], else: asks
            match_orders(new_bids, new_asks, agent_map, tick, nil, vol, trades)
          else
            # Execute trade
            updated_buyer = %{
              buyer
              | gold_inventory: buyer.gold_inventory - match_volume * match_price,
                food_inventory: buyer.food_inventory + match_volume
            }

            updated_seller = %{
              seller
              | gold_inventory: seller.gold_inventory + match_volume * match_price,
                food_inventory: seller.food_inventory - match_volume
            }

            agent_map =
              agent_map
              |> Map.put(buyer.id, updated_buyer)
              |> Map.put(seller.id, updated_seller)

            trade = %TradeEvent{
              tick: tick,
              buyer_id: buyer.id,
              seller_id: seller.id,
              price: match_price,
              qty: match_volume
            }

            new_bid = %{bid | qty: bid.qty - match_volume}
            new_ask = %{ask | qty: ask.qty - match_volume}

            new_bids = if new_bid.qty <= 0, do: rest_bids, else: [new_bid | rest_bids]
            new_asks = if new_ask.qty <= 0, do: rest_asks, else: [new_ask | rest_asks]

            match_orders(
              new_bids,
              new_asks,
              agent_map,
              tick,
              match_price,
              vol + match_volume,
              [trade | trades]
            )
          end
        end
    end
  end
end
